internal/controller/certificaterequests/apply_test.go

```diff
@@ -0,0 +1,128 @@
+// preparation for a Certificate Apply call. Only object meta/type and spec
+// field should be present.
+func Test_serializeApply(t *testing.T) {
+	// Expected serialized Certificate Apply object. Should only contain base
+	// type object, object meta object, and spec. status should be empty. Empty
+	// spec should be deterministic.
+	const (
+		expReg      = `^{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{.*},"spec":{.*},"status":{}}$`
+		expEmptyReg = `^{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{.*},"spec":{"issuerRef":{"name":""},"request":null},"status":{}}$`
+		numJobs     = 10000
+	)
+
+	var wg sync.WaitGroup
+	jobs := make(chan int)
+
+	wg.Add(numJobs)
+	for i := 0; i < 3; i++ {
+		go func() {
+			for j := range jobs {
+				t.Run("fuzz_"+strconv.Itoa(j), func(t *testing.T) {
+					var req cmapi.CertificateRequest
+					fuzz.New().NilChance(0.5).Fuzz(&req)
+
+					// Test regex with non-empty spec.
+					reqData, err := serializeApply(&req)
+					assert.NoError(t, err)
+					assert.Regexp(t, expReg, string(reqData))
+
+					// String match on empty spec.
+					req.Spec = cmapi.CertificateRequestSpec{}
+					reqData, err = serializeApply(&req)
+					assert.NoError(t, err)
+					assert.Regexp(t, expEmptyReg, string(reqData))
+
+					wg.Done()
+				})
+			}
+		}()
+	}
+
+	for i := 0; i < numJobs; i++ {
+		jobs <- i
+	}
+	close(jobs)
+	wg.Wait()
+}
+
+// This test ensures that when a Certificate object is serialized in
+// preparation for a Certificate status Apply call. Only the required
+// metadata/type fields are present, and only empty spec fields are set. We
+// also ensure that all Certificate status fields are tagged omitempty, and are
+// not serialized if unset.
+func Test_serializeApplyStatus(t *testing.T) {
+	// Expected serialized Certificate Apply object. Should only contain base
+	// meta/type object, empty spec. Status should be matched both via regex, and
+	// when empty.
+	const (
+		expReg   = `^{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{"name":"foo","namespace":"bar","creationTimestamp":null},"spec":{"issuerRef":{"name":""},"request":null},"status":{.*}}$`
+		expEmpty = `{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{"name":"foo","namespace":"bar","creationTimestamp":null},"spec":{"issuerRef":{"name":""},"request":null},"status":{}}`
+		numJobs  = 10000
+	)
+
+	var wg sync.WaitGroup
+	jobs := make(chan int)
+
+	wg.Add(numJobs)
+	for i := 0; i < 3; i++ {
+		go func() {
+			for j := range jobs {
+				t.Run("fuzz_"+strconv.Itoa(j), func(t *testing.T) {
+					var req cmapi.CertificateRequest
+					fuzz.New().NilChance(0.5).Fuzz(&req)
+					req.Name = "foo"
+					req.Namespace = "bar"
+
+					// Test regex with non-empty status.
+					reqData, err := serializeApplyStatus(&req)
+					assert.NoError(t, err)
+					assert.Regexp(t, expReg, string(reqData))
```

Comment:
Thought: similarly to the `serializeApply` test, would it make sense to compare with the unmarshaled status of the fuzzed `req` to see that the status fields were preserved correctly?